Demo运行报错，java.lang.UnsupportedOperationException: Can't convert value at index 6 to color: type=0x1

2022-06-24 14:18:22.307 15467-15467/com.tencent.shadow.sample.host E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.tencent.shadow.sample.host:plugin, PID: 15467
    java.lang.RuntimeException: Unable to start activity ComponentInfo{com.tencent.shadow.sample.host/com.tencent.shadow.sample.plugin.runtime.PluginDefaultProxyActivity}: java.lang.RuntimeException: android.view.InflateException: Binary XML file line #29: Binary XML file line #29: Error inflating class com.jeremyfeinstein.slidingmenu.lib.SlidingMenu
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2896)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2974)
        at android.app.ActivityThread.-wrap11(Unknown Source:0)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1675)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:192)
        at android.app.ActivityThread.main(ActivityThread.java:6753)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:549)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:825)
     Caused by: java.lang.RuntimeException: android.view.InflateException: Binary XML file line #29: Binary XML file line #29: Error inflating class com.jeremyfeinstein.slidingmenu.lib.SlidingMenu
        at com.tencent.shadow.core.loader.delegates.ShadowActivityDelegate.onCreate(ShadowActivityDelegate.kt:159)
        at com.tencent.shadow.core.runtime.container.PluginContainerActivity.onCreate(PluginContainerActivity.java:84)
        at android.app.Activity.performCreate(Activity.java:7113)
        at android.app.Activity.performCreate(Activity.java:7104)
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1220)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2849)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2974) 
        at android.app.ActivityThread.-wrap11(Unknown Source:0) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1675) 
        at android.os.Handler.dispatchMessage(Handler.java:106) 
        at android.os.Looper.loop(Looper.java:192) 
        at android.app.ActivityThread.main(ActivityThread.java:6753) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:549) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:825) 
     Caused by: android.view.InflateException: Binary XML file line #29: Binary XML file line #29: Error inflating class com.jeremyfeinstein.slidingmenu.lib.SlidingMenu
     Caused by: android.view.InflateException: Binary XML file line #29: Error inflating class com.jeremyfeinstein.slidingmenu.lib.SlidingMenu
     Caused by: java.lang.reflect.InvocationTargetException
        at java.lang.reflect.Constructor.newInstance0(Native Method)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:334)
        at android.view.LayoutInflater.createView(LayoutInflater.java:656)
        at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:835)
        at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:775)
        at android.view.LayoutInflater.inflate(LayoutInflater.java:498)
        at android.view.LayoutInflater.inflate(LayoutInflater.java:429)
        at android.view.LayoutInflater.inflate(LayoutInflater.java:380)
        at com.tencent.shadow.core.runtime.ShadowActivity.setContentView(ShadowActivity.java:40)
        at com.tencent.shadow.sample.plugin.app.lib.gallery.MainActivity.onCreate(MainActivity.java:65)
        at com.tencent.shadow.core.loader.delegates.ShadowActivityDelegate.onCreate(ShadowActivityDelegate.kt:156)
        at com.tencent.shadow.core.runtime.container.PluginContainerActivity.onCreate(PluginContainerActivity.java:84)
        at android.app.Activity.performCreate(Activity.java:7113)
        at android.app.Activity.performCreate(Activity.java:7104)
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1220)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2849)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2974)
        at android.app.ActivityThread.-wrap11(Unknown Source:0)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1675)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:192)
        at android.app.ActivityThread.main(ActivityThread.java:6753)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:549)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:825)
2022-06-24 14:18:22.309 15467-15467/com.tencent.shadow.sample.host E/AndroidRuntime: Caused by: android.view.InflateException: Binary XML file line #52: Binary XML file line #52: Error inflating class <unknown>
    Caused by: android.view.InflateException: Binary XML file line #52: Error inflating class <unknown>
    Caused by: java.lang.reflect.InvocationTargetException
        at java.lang.reflect.Constructor.newInstance0(Native Method)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:334)
        at android.view.LayoutInflater.createView(LayoutInflater.java:656)
        at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:835)
        at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:775)
        at android.view.LayoutInflater.rInflate(LayoutInflater.java:908)
        at android.view.LayoutInflater.rInflateChildren(LayoutInflater.java:869)
        at android.view.LayoutInflater.rInflate(LayoutInflater.java:911)
        at android.view.LayoutInflater.rInflateChildren(LayoutInflater.java:869)
        at android.view.LayoutInflater.inflate(LayoutInflater.java:521)
        at android.view.LayoutInflater.inflate(LayoutInflater.java:429)
        at android.view.LayoutInflater.inflate(LayoutInflater.java:380)
        at com.jeremyfeinstein.slidingmenu.lib.SlidingMenu.setContent(SlidingMenu.java:353)
        at com.jeremyfeinstein.slidingmenu.lib.SlidingMenu.<init>(SlidingMenu.java:246)
        at com.jeremyfeinstein.slidingmenu.lib.SlidingMenu.<init>(SlidingMenu.java:197)
        at java.lang.reflect.Constructor.newInstance0(Native Method)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:334)
        at android.view.LayoutInflater.createView(LayoutInflater.java:656)
        at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:835)
        at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:775)
        at android.view.LayoutInflater.inflate(LayoutInflater.java:498)
        at android.view.LayoutInflater.inflate(LayoutInflater.java:429)
        at android.view.LayoutInflater.inflate(LayoutInflater.java:380)
        at com.tencent.shadow.core.runtime.ShadowActivity.setContentView(ShadowActivity.java:40)
        at com.tencent.shadow.sample.plugin.app.lib.gallery.MainActivity.onCreate(MainActivity.java:65)
        at com.tencent.shadow.core.loader.delegates.ShadowActivityDelegate.onCreate(ShadowActivityDelegate.kt:156)
        at com.tencent.shadow.core.runtime.container.PluginContainerActivity.onCreate(PluginContainerActivity.java:84)
        at android.app.Activity.performCreate(Activity.java:7113)
        at android.app.Activity.performCreate(Activity.java:7104)
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1220)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2849)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2974)
        at android.app.ActivityThread.-wrap11(Unknown Source:0)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1675)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:192)
        at android.app.ActivityThread.main(ActivityThread.java:6753)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:549)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:825)
     Caused by: java.lang.UnsupportedOperationException: Can't convert value at index 6 to color: type=0x1
        at android.content.res.TypedArray.getColor(TypedArray.java:473)
        at android.widget.AbsListView.<init>(AbsListView.java:894)
        at android.widget.ListView.<init>(ListView.java:227)
        at android.widget.ExpandableListView.<init>(ExpandableListView.java:234)
        at android.widget.ExpandableListView.<init>(ExpandableListView.java:229)
        at android.widget.ExpandableListView.<init>(ExpandableListView.java:225)
        at com.ryg.expandable.ui.PinnedHeaderExpandableListView.<init>(PinnedHeaderExpandableListView.java:74)
        at java.lang.reflect.Constructor.newInstance0(Native Method) 
        at java.lang.reflect.Constructor.newInstance(Constructor.java:334) 
        at android.view.LayoutInflater.createView(LayoutInflater.java:656) 
        at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:835) 
        at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:775) 
        at android.view.LayoutInflater.rInflate(LayoutInflater.java:908) 
        at android.view.LayoutInflater.rInflateChildren(LayoutInflater.java:869) 
        at android.view.LayoutInflater.rInflate(LayoutInflater.java:911) 
        at android.view.LayoutInflater.rInflateChildren(LayoutInflater.java:869) 
        at android.view.LayoutInflater.inflate(LayoutInflater.java:521) 
        at android.view.LayoutInflater.inflate(LayoutInflater.java:429) 
        at android.view.LayoutInflater.inflate(LayoutInflater.java:380) 
        at com.jeremyfeinstein.slidingmenu.lib.SlidingMenu.setContent(SlidingMenu.java:353) 
        at com.jeremyfeinstein.slidingmenu.lib.SlidingMenu.<init>(SlidingMenu.java:246) 
        at com.jeremyfeinstein.slidingmenu.lib.SlidingMenu.<init>(SlidingMenu.java:197) 

> 看起来可能是API 27以下对于资源ID分区的实现还是有bug。可能是那个getColor有问题，得再构造下测试用例看看。
> 
> 不过肯定是要把使用MixResource方案的最高API提升到27了。

@shifujun 

Demo崩溃的问题解决了；
但是我尝试使用AndroidX中的AppCompatActivity，却遇到一个崩溃问题：
Caused by: java.lang.IllegalStateException: You need to use a Theme.AppCompat theme (or descendant) with this activity.
        at androidx.appcompat.app.AppCompatDelegateImpl.createSubDecor(AppCompatDelegateImpl.java:852)
        at androidx.appcompat.app.AppCompatDelegateImpl.ensureSubDecor(AppCompatDelegateImpl.java:815)
        at androidx.appcompat.app.AppCompatDelegateImpl.setContentView(AppCompatDelegateImpl.java:703)
        at androidx.appcompat.app.AppCompatActivity.setContentView(AppCompatActivity.java:195)
在插件Activity启动的时候无法读取到androidx.appcompat.R.styleable.AppCompatTheme主题中的androidx.appcompat.R.styleable.AppCompatTheme_windowActionBar属性，导致启动的时候Android系统认定主题没有使用AppCompat（实际上已经使用）

然后我在插件中使用如下代码测试：
![image](https://user-images.githubusercontent.com/17923307/175554830-d4ae24a6-6d00-42db-aa31-ab382f1a2f23.png)
打印的结果是false

然后我使用shadow工程中的demo进行了如下的测试：
1、在Host中增加了一个自定义属性，并在在主题中设置该属性。让Host的Application使用该主题
![image](https://user-images.githubusercontent.com/17923307/175559603-c98e11b1-2db4-4006-a2a6-2ab4990b11ac.png)
2、在Host的MainActivity中测试：
测试代码：
` String name = getResources().getResourceName(com.tencent.shadow.sample.host.lib.R.layout.host_ui_layer_layout);
        Resources.Theme t = getTheme();
        t.applyStyle(com.tencent.shadow.sample.host.lib.R.style.TestHostTheme, true);
        TypedArray ta = t.obtainStyledAttributes(com.tencent.shadow.sample.host.lib.R.styleable.DominoScroll);
        boolean re = ta.hasValue(com.tencent.shadow.sample.host.lib.R.styleable.DominoScroll_dominoContent);
        ta.recycle();
        Log.i("ttest", "host hasValue= " + re + ",name="+name);`
![image](https://user-images.githubusercontent.com/17923307/175559919-bc9b3629-29bb-4157-b07a-12a44041aad7.png)
能够获取到自定义的属性
3、在Plugin的MainActivity中测试：(这里将Host用到的主题和属性都移到了Host-lib中，使plugin能够正常访问)
测试代码：
` String name = getResources().getResourceName(com.tencent.shadow.sample.host.lib.R.layout.host_ui_layer_layout);
        Resources.Theme t = getTheme();
        t.applyStyle(com.tencent.shadow.sample.host.lib.R.style.TestHostTheme, true);
        TypedArray ta = t.obtainStyledAttributes(com.tencent.shadow.sample.host.lib.R.styleable.DominoScroll);
        boolean re = ta.hasValue(com.tencent.shadow.sample.host.lib.R.styleable.DominoScroll_dominoContent);

        ta.recycle();
        Log.i("ttest", "plugin hasValue= " + re + ",name=" +name);`
![image](https://user-images.githubusercontent.com/17923307/175560615-b6ba0828-2e21-4781-89e6-0f95c5125b0a.png)

经过上面三步测试，发现一个问题:
插件的拿到的Resource主题无法获取宿主中的属性，从而导致当启动androidX的AppCOmpatActivity‘时候无法获取’androidx.appcompat.R.styleable.AppCompatTheme_windowActionBar属性导致出现上述崩溃；
这个崩溃问题在API 31上没有出现，但是在API27上出现了
想问下是什么原因呢

--- a/projects/sdk/core/loader/src/main/kotlin/com/tencent/shadow/core/loader/blocs/CreateResourceBloc.kt
+++ b/projects/sdk/core/loader/src/main/kotlin/com/tencent/shadow/core/loader/blocs/CreateResourceBloc.kt
@@ -36,6 +36,21 @@ import android.webkit.WebView
 import java.util.concurrent.CountDownLatch
 
 object CreateResourceBloc {
+    /**
+     * 构造插件Resources时有两个方案，都要解决一个共同的问题：
+     * 系统有可能从宿主Manifest中获取app icon或者logo的资源ID，
+     * 然后直接向插件的Resources对象查询这些资源。
+     *
+     * 第一个方案是MixResources方案，但该方案依赖Resources的Deprecated构造器，
+     * 未来可能会不可用。实际上Resources的构造器如果不取消的话，这个方案可以一直使用下去。
+     *
+     * 第二个方案是利用资源分区，这是一个和AAB设计中dynamic-feature相同的方案，
+     * 将宿主和插件apk添加到同一个Resources对象中。
+     * 尽管构造这种带有多资源ID分区的Resources对象所需的API在低版本系统上就已经有了，
+     * 但通过不断测试发现MAX_API_FOR_MIX_RESOURCES及更低的API系统上，有个别API不能正确支持非0x7f分区的资源。
+     */
+    const val MAX_API_FOR_MIX_RESOURCES = Build.VERSION_CODES.O_MR1
+
     fun create(archiveFilePath: String, hostAppContext: Context): Resources {
         triggerWebViewHookResources(hostAppContext)
 
@@ -45,7 +60,7 @@ object CreateResourceBloc {
         applicationInfo.packageName = hostApplicationInfo.packageName
         applicationInfo.uid = hostApplicationInfo.uid
 
-        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+        if (Build.VERSION.SDK_INT > MAX_API_FOR_MIX_RESOURCES) {
             fillApplicationInfoForNewerApi(applicationInfo, hostApplicationInfo, archiveFilePath)
         } else {
             fillApplicationInfoForLowerApi(applicationInfo, hostApplicationInfo, archiveFilePath)
@@ -54,7 +69,7 @@ object CreateResourceBloc {
         try {
             val pluginResource = packageManager.getResourcesForApplication(applicationInfo)
 
-            return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+            return if (Build.VERSION.SDK_INT > MAX_API_FOR_MIX_RESOURCES) {
                 pluginResource
             } else {
                 val hostResources = hostAppContext.resources
@@ -142,7 +157,7 @@ object CreateResourceBloc {
  * 因调用addAssetPath方法也无法满足CreateResourceTest涉及的场景。
  */
 @Suppress("DEPRECATION", "OVERRIDE_DEPRECATION")
-@TargetApi(25)
+@TargetApi(CreateResourceBloc.MAX_API_FOR_MIX_RESOURCES)
 private class MixResources(
     private val mainResources: Resources,
     private val sharedResources: Resources

--- a/projects/test/dynamic/host/test-dynamic-host/src/androidTest/java/com/tencent/shadow/test/cases/plugin_main/NullRefInXmlTest.java
+++ b/projects/test/dynamic/host/test-dynamic-host/src/androidTest/java/com/tencent/shadow/test/cases/plugin_main/NullRefInXmlTest.java
@@ -1,0 +1,49 @@
+/*
+ * Tencent is pleased to support the open source community by making Tencent Shadow available.
+ * Copyright (C) 2019 THL A29 Limited, a Tencent company.  All rights reserved.
+ *
+ * Licensed under the BSD 3-Clause License (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *     https://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package com.tencent.shadow.test.cases.plugin_main;
+
+import android.content.Intent;
+
+import androidx.test.core.app.ApplicationProvider;
+import androidx.test.ext.junit.runners.AndroidJUnit4;
+
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+
+@RunWith(AndroidJUnit4.class)
+public class NullRefInXmlTest extends PluginMainAppTest {
+
+    @Override
+    protected Intent getLaunchIntent() {
+        Intent pluginIntent = new Intent();
+        String packageName = ApplicationProvider.getApplicationContext().getPackageName();
+        pluginIntent.setClassName(
+                packageName,
+                "com.tencent.shadow.test.plugin.general_cases.lib.usecases.view.TestNullRefInXmlActivity"
+        );
+        return pluginIntent;
+    }
+
+    @Test
+    public void testColorOfNull() {
+        matchTextWithViewTag("cacheColorHint", "0");
+    }
+
+}

--- a/projects/test/plugin/general-cases/test-plugin-general-cases/src/main/AndroidManifest.xml
+++ b/projects/test/plugin/general-cases/test-plugin-general-cases/src/main/AndroidManifest.xml
@@ -99,6 +99,7 @@
         <activity android:name=".lib.usecases.context.TestLayoutInflaterActivity" />
         <activity android:name=".lib.usecases.context.ContextGetPackageCodePathTestActivity" />
         <activity android:name=".lib.usecases.view.TestViewIdActivity" />
+        <activity android:name=".lib.usecases.view.TestNullRefInXmlActivity" />
 
         <provider
             android:name=".lib.usecases.provider.TestProvider"

--- a/projects/test/plugin/general-cases/test-plugin-general-cases/src/main/java/com/tencent/shadow/test/plugin/general_cases/lib/usecases/view/TestNullRefInXmlActivity.java
+++ b/projects/test/plugin/general-cases/test-plugin-general-cases/src/main/java/com/tencent/shadow/test/plugin/general_cases/lib/usecases/view/TestNullRefInXmlActivity.java
@@ -1,0 +1,34 @@
+package com.tencent.shadow.test.plugin.general_cases.lib.usecases.view;
+
+import android.app.Activity;
+import android.os.Bundle;
+import android.support.annotation.Nullable;
+import android.view.ViewGroup;
+import android.widget.ListView;
+
+import com.tencent.shadow.test.plugin.general_cases.R;
+import com.tencent.shadow.test.plugin.general_cases.lib.gallery.util.UiUtil;
+
+public class TestNullRefInXmlActivity extends Activity {
+
+    private ViewGroup mItemViewGroup;
+
+    @Override
+    protected void onCreate(@Nullable Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+
+        ListView listView = (ListView) getLayoutInflater().inflate(R.layout.layout_test_null_ref, null);
+        int cacheColorHint = listView.getCacheColorHint();
+
+        mItemViewGroup = UiUtil.setActivityContentView(this);
+
+        mItemViewGroup.addView(
+                UiUtil.makeItem(
+                        this,
+                        "cacheColorHint",
+                        "cacheColorHint",
+                        Integer.toString(cacheColorHint, 16)
+                )
+        );
+    }
+}

--- a/projects/test/plugin/general-cases/test-plugin-general-cases/src/main/res/layout/layout_test_null_ref.xml
+++ b/projects/test/plugin/general-cases/test-plugin-general-cases/src/main/res/layout/layout_test_null_ref.xml
@@ -1,0 +1,5 @@
+<?xml version="1.0" encoding="utf-8"?>
+<ListView xmlns:android="http://schemas.android.com/apk/res/android"
+    android:layout_height="match_parent"
+    android:layout_width="match_parent"
+    android:cacheColorHint="@null" />

